drivers/hap_report/app/models/hap_report/report.rb

```diff
@@ -0,0 +1,325 @@
+###
+# Copyright 2016 - 2021 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+module HapReport
+  class Report < SimpleReports::ReportInstance
+    include Queries
+    after_initialize :set_attributes
+
+    HAP_FUNDING = 'HAP Funded'.freeze
+
+    attr_accessor :start_date, :end_date, :project_ids
+    validates_presence_of :start_date, :end_date, :project_ids
+
+    def build_report
+      update(status: :processing)
+      create_universe
+      report_labels.values.each do |sections|
+        sections.values.each do |row_scope|
+          project_columns.values.each do |column_scope|
+            cell_name = "#{row_scope}_#{column_scope}"
+            next if blank_cells.include?(cell_name)
+
+            cell = report_cells.create(name: cell_name)
+            if row_scope == :total_units_of_shelter_service
+              cell.summary = report_client_scope.where(send(column_scope)).sum(a_t[:nights_in_shelter])
+            else
+              cell_scope = send(row_scope).where(send(column_scope))
+              cell.add_members(cell_scope)
+              cell.summary = cell_scope.count
+            end
+            cell.save!
+          end
+        end
+      end
+      update(status: :completed)
+    end
+
+    private def create_universe
+      hap_clients = {}
+      enrollment_scope.find_each do |processed_enrollment|
```

Comment:
Given the size of the preloads, we may need to limit the batch size here.  I'm unclear on how many clients each report may be looking at at this time.